src/components/study-code-upload.tsx

```diff
@@ -31,40 +32,76 @@ import { uniqueBy } from 'remeda'
 import { LaunchIDEButton, OrDivider, UploadFilesButton } from './study/study-upload-buttons'
 
 interface StudyCodeUploadProps {
-    studyUploadForm: UseFormReturnType<StudyJobCodeFilesValues>
-    showStepIndicator?: boolean
+    studyUploadForm?: UseFormReturnType<StudyJobCodeFilesValues> | null
+    stepIndicator?: string
     title?: string
     language: Language
     orgSlug: string
+    studyId?: string | null
+    onBeforeLaunchIDE?: () => Promise<string | null> // Returns studyId after saving draft, or null if failed
+    onIDELaunched?: () => void // Called when IDE launch is initiated
+    onIDELoadingChange?: (loading: boolean) => void // Called when IDE loading state changes
 }
 
 export const StudyCodeUpload = ({
-    showStepIndicator = false,
+    stepIndicator,
     title = 'Study code',
     language,
     orgSlug,
+    studyId,
+    onBeforeLaunchIDE,
+    onIDELaunched,
+    onIDELoadingChange,
     studyUploadForm,
 }: StudyCodeUploadProps) => {
     const [isModalOpen, { open: openModal, close: closeModal }] = useDisclosure(false)
     const [isAlertVisible, setIsAlertVisible] = useState(true)
+    const [effectiveStudyId, setEffectiveStudyId] = useState<string | null>(studyId ?? null)
     const theme = useMantineTheme()
 
+    const {
+        launchWorkspace,
+        isLaunching: isLaunchingWorkspace,
+        isCreatingWorkspace,
+        error: launchError,
+    } = useWorkspaceLauncher({
+        studyId: effectiveStudyId ?? '',
```

Comment:
If this is used before a draft is saved, we can end up calling createUserAndWorkspaceAction with an empty studyid (not allowed). I see onBeforeLaunchIDE as a kind of safeguard here below. Is that safe enough?

Should there be more safeguards inside useWorkspaceLauncher?